### DISCUSSION TITLE
RemoteControl HTTP API - move function

### DISCUSSION THREAD

**Discussion by zac-powers** [Stellarium/stellarium#3788] [Q&A] [answered]
For context, I am trying to use Stellarium to test some satellite tracking software, so I need to be able to control azimuth, elevation, azimuth rate, and elevation rate. I am trying to manually control the view speed using [http requests](https://stellarium.org/doc/0.20/remoteControlApi.html#rcScriptServicePOST) through the remote control plugin. The move() function accepts an x value and y value the controls the direction and speed in the azimuth and elevation directions.

My problem is that this speed seems to be defined in a weird way. It seems like it is dependent upon the fov. For example if you are zoomed out to a wide fov, using move(1,1) will move the view very fast, while if you are zoomed in to a small fov, the move speed will be slow. This behavior aligns with how using the arrow keys works, which I assume is what the function was built for. But for my purposes, I need to be able to input an azimuth or elevation rate in deg/sec, keeping a constant movement rate independent of the fov. Do you know if this is possible? 

If its not possible, there should be some theoretical fov such that using move(1,1) coincidentally corresponds with 
move(1 deg/sec, 1deg/sec). Would there be an easy way to find this besides checking an insane amount of fovs and doing a bunch of tedious calculations with pixel widths and view angles? 

Here is my function that formats the http request. It is written using python, the for loop just moves for a certain amount of time.

```
def move(az_r, el_r, sec):
    for x in range(1,sec):
        a = {'x' : az_r, 'y' : el_r}
        r = requests.post("http://localhost:8090/api/main/move?", a)
        time.sleep(1)
```

I suppose another way to ask the question would be, for example if you click on a satellite and center on screen, what is Stellarium doing behind the scenes that controls the view, and is there a way to manually implement this using the http API tools.

**Comment by gzotti** [comment]
I am afraid you are on you own to run your tests here. You can center the screen on objects like satellites. Those find their positions and center screen accordingly. But the RC plugin was not developed for things you want, but those that you assumed. See the regular HTTP web interface.

**Reply by zac-powers** [reply]
Ok. I would think that since the speed is dependent on the fov, there would be some scale factor between the fov and the change in view angle. Do you know how Stellarium handles this, or where in the source code would be a good place to look? Otherwise I will just begin doing some trial and error.

**Reply by gzotti** [reply]
StelMovementMgr. Good Luck!

**Comment by zac-powers** [comment, marked as answer]
For anyone else, I found what I think is the answer in stellarium/plugins/RemoteControl/src/MainService.cpp line 88

```
double currentFov = mvmgr->getCurrentFov();
// the more it is zoomed, the lower the moving speed is (in angle)
//0.0004 is the default key move speed
double depl=0.0004 / 30 *deltaTime*1000*currentFov;

double deltaAz = moveX*depl;
double deltaAlt = moveY*depl;

mvmgr->panView(deltaAz,deltaAlt);
```

It still seemed a little off, so I adjusted it slightly, but this code just about does the trick, works best ~2deg fov, but obviously can be tuned slightly. Also I'm not 100% confident in the precision of my az/el rates, so take that with a grain of salt, the scaler may need to be adjusted some more (astropy seems to have a little trouble converting speed to Azimuth/Elevation).
```
def move(az, el, sec, fov_num):
    fov(fov_num)
    for x in range(1,sec):
        scaler=0.00047/fov_num*1000*3
        a = {'x' : az*scaler, 'y' : el*scaler}
        r = requests.post("http://localhost:8090/api/main/move?", a)
        time.sleep(1))
```
I can post the full code in the future if anyone is curious on how to make their own telescope simulator. I currently have it working to where I can put in a TLE and it will track it pretty closely.
